src/core/unit.c

```diff
@@ -4256,6 +4262,53 @@ static int unit_verify_contexts(const Unit *u) {
         return 0;
 }
 
+static PrivateTmp unit_get_private_var_tmp(Unit *u, ExecContext *c) {
+        assert(u);
+        assert(c);
+        assert(c->private_tmp >= 0 && c->private_tmp < _PRIVATE_TMP_MAX);
+
+        /* Disable disconnected private tmpfs on /var/tmp/ when DefaultDependencies=no and
+         * RootImage=/RootDirectory= are not set, as /var/ may be a separated partition.
+         * See issue #37258. */
+
+        /* PrivateTmp=yes/no also enables/disables private tmpfs on /var/tmp/. */
+        if (c->private_tmp != PRIVATE_TMP_DISCONNECTED)
+                return c->private_tmp;
+
+        /* When DefaultDependencies=yes, disconnected tmpfs is also enabled on /var/tmp/, and an explicit
+         * dependency to the mount on /var/ will be added in unit_add_exec_dependencies(). */
+        if (u->default_dependencies)
+                return PRIVATE_TMP_DISCONNECTED;
+
+        /* When RootImage=/RootDirectory= is enabled, /var/ should be prepared by the image or directory,
+         * hence we can mount a disconnected tmpfs on /var/tmp/. */
+        if (exec_context_with_rootfs(c))
+                return PRIVATE_TMP_DISCONNECTED;
+
+        /* Even if DefaultDependencies=no, enable disconnected tmpfs when
+         * RequiresMountsFor=WantsMountsFor=/var/ is explicitly set. */
+        for (UnitMountDependencyType t = 0; t < _UNIT_MOUNT_DEPENDENCY_TYPE_MAX; t++)
+                if (hashmap_contains(u->mounts_for[t], "/var/"))
+                        return PRIVATE_TMP_DISCONNECTED;
+
+        /* Check the same but for After= and Requires=/Wants=. */
+        Unit *m = manager_get_unit(u->manager, "var.mount");
+        if (!m)
+                return PRIVATE_TMP_NO;
+
+        if (!unit_has_dependency(u, UNIT_ATOM_AFTER, m))
+                return PRIVATE_TMP_NO;
+
+        if (unit_has_dependency(u, unit_dependency_to_atom(UNIT_REQUIRES), m))
+                return PRIVATE_TMP_DISCONNECTED;
+
+        if (unit_has_dependency(u, unit_dependency_to_atom(UNIT_WANTS), m))
+                return PRIVATE_TMP_DISCONNECTED;
```

Comment:
hmm, i'd probably not bother with catching Wants=/Requires= here. Any setup with `var.mount` around and carries `After=var.mount` for the unit is 99.99% going to have `/var/` mounted by local-fs.target. The rest 0.01% is outright broken, because `/var/` will be pulled in by any unit with StateDirectory= and friends or PrivateTmp=yes. We don't need to care about them if you ask me.

Reply:
and there's BindsTo=. So the current list is non-exhaustive

Reply:
But, RequiresMountsFor= is exactly the same as After= and Requires= especially for /var/. So, I do not think we should take one of them. If we take RequiresMountsFor=, we should also take After= and Requires=.

There is no BindsMountsFor=. So, I think it is OK to not consider BindsTo=, even though it should work. People can specify both BindsTo=var.mount and RequiresMountsFor=/var/ at the same time.

Reply:
> But, RequiresMountsFor= is exactly the same as After= and Requires= especially for /var/. So, I do not think we should take one of them. If we take RequiresMountsFor=, we should also take After= and Requires=.

They aren't though. `Requires=` mandates var.mount to exist, while mounts for deps first checks if it's a mountpoint at all. IOW, `RequiresMountsFor=` is inherently weaker than Requires=, and it tries to ensure /var/ is around, not about var.mount being _mounted_.

(a quick search reveals several occurrences of standalone `After=var.mount` btw: https://grep.app/search?case=true&q=After%3Dvar.mount , exactly because of the reasons stated above)

Reply:
I mean, they probably come from the ancient era before RequiresMountsFor=. But the fact that `After=var.mount` exists kinda suggests their expectation on `/var/`. So we can just enable `/var/` for them...

Reply:
Even though, the net effect for the current problem should be the same. If a unit has Requires= and After= var.mount, then we can safely mount tmpfs on /var/tmp/.

Reply:
With After=var.mount it should be safe too... Or, please check for UNIT_ATOM_PULL_IN_START and UNIT_ATOM_PULL_IN_START_IGNORED. I really think Requires= and Wants= is not special here.

Reply:
Done. PTAL.